src/devices/src/virtio/net/device.rs

```diff
@@ -409,24 +425,28 @@ impl Net {
     fn write_to_mmds_or_tap(
         mmds_ns: Option<&mut MmdsNetworkStack>,
         rate_limiter: &mut RateLimiter,
-        frame_buf: &[u8],
+        frame_buf: &mut [u8],
+        frame_iovec: &mut IoVecBuffer,
         tap: &mut Tap,
         guest_mac: Option<MacAddr>,
     ) -> Result<bool> {
-        let checked_frame = |frame_buf| {
-            frame_bytes_from_buf(frame_buf).map_err(|e| {
-                error!("VNET header missing in the TX frame.");
-                METRICS.net.tx_malformed_frames.inc();
-                e
-            })
-        };
+        let frame_header = frame_bytes_from_buf(frame_buf).map_err(|e| {
+            error!("VNET header missing in the TX frame.");
+            METRICS.net.tx_malformed_frames.inc();
+            e
+        })?;
+
         if let Some(ns) = mmds_ns {
-            if ns.detour_frame(checked_frame(frame_buf)?) {
+            if ns.is_mmds_frame(frame_header) {
+                // Here it is ok to unwrap. We have already checked that the IoVecBuffer holds
+                // at least the frame headers and we are reading from the beginning of
+                // the IoVecBuffer.
+                let frame_len = frame_iovec.read_at(frame_buf, 0).unwrap();
+                let _ = ns.detour_frame(&frame_buf[..frame_len]);
```

Comment:
This looks wrong.. it sends entire `frame_iovec` contents to `mmds.detour_frame()`, including prepended VNET header. I believe it should strip the VNET header before.

Also since MMDS frames are not on the critical path you can use a stack-allocated buffer here for them instead of the `self.frame_buf` cache. This would also allow `self.frame_buf` to be reduced in size from `65k` to a few hundred bytes required by headers.

```suggestion
                let mut mmds_frame = [0u8; frame_iovec.len()];
                let frame_len = frame_iovec.read_at(&mut mmds_frame, vnet_hdr_len()).unwrap();
                let _ = ns.detour_frame(&mmds_frame[..frame_len]);
```

Reply:
I think you are right. I haven't looked into the mmds path very closely yet. I 'll look into it.